src/main/java/com/linecorp/armeria/common/http/Http1ClientUpgradeHandler.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2014 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.linecorp.armeria.common.http;
+
+import static io.netty.handler.codec.http.HttpResponseStatus.SWITCHING_PROTOCOLS;
+import static io.netty.util.ReferenceCountUtil.release;
+
+import java.net.SocketAddress;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandler;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpClientUpgradeHandler.UpgradeCodec;
+import io.netty.handler.codec.http.HttpClientUpgradeHandler.UpgradeEvent;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpObject;
+import io.netty.handler.codec.http.HttpObjectAggregator;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.util.AsciiString;
+
+/**
+ * Client-side handler for handling an HTTP upgrade handshake to another protocol. When the first
+ * HTTP request is sent, this handler will add all appropriate headers to perform an upgrade to the
+ * new protocol. If the upgrade fails (i.e. response is not 101 Switching Protocols), this handler
+ * simply removes itself from the pipeline. If the upgrade is successful, upgrades the pipeline to
+ * the new protocol.
+ */
+public class Http1ClientUpgradeHandler extends HttpObjectAggregator implements ChannelOutboundHandler {
+
+    /**
+     * The source codec that is used in the pipeline initially.
+     */
+    public interface SourceCodec {
+
+        /**
+         * Removes or disables the encoder of this codec so that the {@link UpgradeCodec} can send an initial greeting
+         * (if any).
+         */
+        void prepareUpgradeFrom(ChannelHandlerContext ctx);
+
+        /**
+         * Removes this codec (i.e. all associated handlers) from the pipeline.
+         */
+        void upgradeFrom(ChannelHandlerContext ctx);
+    }
+
+    private final SourceCodec sourceCodec;
+    private final UpgradeCodec upgradeCodec;
+    private boolean upgradeRequested;
```

Comment:
:+1:
On Feb 2, 2016 7:54 PM, "Trustin Lee" <notifications@github.com> wrote:

> In
> src/main/java/com/linecorp/armeria/common/http/Http1ClientUpgradeHandler.java
> <https://github.com/line/armeria/pull/102#discussion_r51552856>:
>
> > +
> > +        /**
> > +         * Removes or disables the encoder of this codec so that the {@link UpgradeCodec} can send an initial greeting
> > +         * (if any).
> > +         */
> > +        void prepareUpgradeFrom(ChannelHandlerContext ctx);
> > +
> > +        /**
> > +         * Removes this codec (i.e. all associated handlers) from the pipeline.
> > +         */
> > +        void upgradeFrom(ChannelHandlerContext ctx);
> > +    }
> > +
> > +    private final SourceCodec sourceCodec;
> > +    private final UpgradeCodec upgradeCodec;
> > +    private boolean upgradeRequested;
>
> Yes. Netty handler methods are invoked from the same thread.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/line/armeria/pull/102/files#r51552856>.
>